design-documents/promises.md

```diff
@@ -0,0 +1,112 @@
+### Why?
+Service contracts in the future will often be executed in an asynchronous manner
+and it's time to introduce a standard Promise to Magento for asynchronous operations to employ
+### Requirements
+* Promises CANNOT be forwarded with _then_ and _otherwise_ ([see explanation](#forwarding))
+* _then_ accepts a function that will be executed when the promise is resolved, the callback will
```

Comment:
When promise is resolved? When we received response on our request or in case of async operation we retried result of our operation? How do we retrieve result of our operation? Can we have API that would allow to create multiple promises and execute code only when all of them resolved?

Reply:
you will be retreiving promised results as the 1st argument to _then_ callback. Creating an util for waiting for multiple promises doesn't seem necessary - you can just use _wait_ on promises you want to wait for